### ISSUE TITLE
Null check not detected after a switch

### ISSUE THREAD

**Issue by BenoitDuffez** [vimeo/psalm#4493]
I'm not sure this is a bug but I don't understand this:

    public function isValid(?string $input): bool {
        switch ($input) {
        case null:
            return false;
        case 'whatever':
            return true;
        }
    
        return preg_match('/[a-z]+/', $input) !== false;
    }

Result:

    psalm: PossiblyNullArgument: Argument 2 of preg_match cannot be null, possibly null value provided

I don't see how it could be null?
Using psalm 4.1.1 on php 7.4

**Comment by BenoitDuffez** [comment]
Of course mr bot. https://psalm.dev/r/1b536a50c7

**Comment by weirdan** [comment]
After `switch ($input) { case null: return false; }` `$input` should be inferred as `non-empty-string` (because it would also bail out on `string('')`, but Psalm thinks it's `null|string`.

**Comment by muglug** [comment]
Simpler reproducer

```php
class A {}

function isValid(?A $input): A {
    if ($input === null) {
        return new A();
    }

    return $input;
}
```

**Comment by orklah** [comment]
I'm gonna close this as a duplicate of https://github.com/vimeo/psalm/issues/3373
